File: backend/app/routers/delivery_notes.py
"""Delivery Note CRUD router with confirm ISSUE / cancel reverse."""

from typing import Optional
from uuid import UUID

from fastapi import APIRouter, Body, Depends, HTTPException, Query, status
from sqlalchemy import func, select
from sqlmodel.ext.asyncio.session import AsyncSession

from app.auth.dependencies import get_current_user, get_current_workspace_id
from app.database import get_session
from app.models.delivery_note import DeliveryNote, DeliveryNoteStatus
from app.models.user import User
from app.schemas.common import PaginatedResponse, PaginationMeta, SuccessResponse
from app.schemas.delivery_notes import (
    DeliveryNoteCancelRequest,
    DeliveryNoteConfirmRequest,
    DeliveryNoteCreate,
    DeliveryNoteResponse,
    DeliveryNoteUpdate,
)
from app.services.delivery_note_service import DeliveryNoteService

router = APIRouter(prefix="/delivery-notes", tags=["Delivery Notes"])


async def _load_or_404(
    session: AsyncSession,
    dn_id: UUID,
    workspace_id: UUID,
    *,
    for_update: bool = False,
) -> DeliveryNote:
    dn = await DeliveryNoteService.get_visible(
        session, dn_id, workspace_id, for_update=for_update
    )
    if dn is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Delivery note not found"
        )
    return dn


async def _wrapped(
    session: AsyncSession, dn: DeliveryNote
) -> SuccessResponse[DeliveryNoteResponse]:
    return SuccessResponse(data=await DeliveryNoteService.wrapped(session, dn))


@router.post(
    "",
    response_model=SuccessResponse[DeliveryNoteResponse],
    status_code=status.HTTP_201_CREATED,
)
async def create_delivery_note(
    body: DeliveryNoteCreate,
    session: AsyncSession = Depends(get_session),
    user: User = Depends(get_current_user),
    workspace_id: UUID = Depends(get_current_workspace_id),
):
    """Create a DRAFT delivery note with DN-YYYY-XXXX numbering."""
    dn = await DeliveryNoteService.create(
        session=session,
        workspace_id=workspace_id,
        user_id=user.id,
        warehouse_id=body.warehouse_id,
        customer_purchase_order_id=body.customer_purchase_order_id,
        invoice_id=body.invoice_id,
        bin_id=body.bin_id,
        delivery_date=body.delivery_date,
        shipping_address=body.shipping_address,
        vehicle_number=body.vehicle_number,
        driver_name=body.driver_name,
        notes=body.notes,
        items=[item.model_dump() for item in body.items] if body.items else None,
    )
    await session.commit()
    return await _wrapped(session, dn)


@router.get("", response_model=PaginatedResponse)
async def list_delivery_notes(
    status_filter: Optional[DeliveryNoteStatus] = Query(None, alias="status"),
    client_id: Optional[UUID] = Query(None),
    customer_purchase_order_id: Optional[UUID] = Query(None),
    invoice_id: Optional[UUID] = Query(None),
    search: Optional[str] = Query(None, description="Search DN number"),
    page: int = Query(1, ge=1),
    per_page: int = Query(20, ge=1, le=100),
    session: AsyncSession = Depends(get_session),
    workspace_id: UUID = Depends(get_current_workspace_id),
):
    """List delivery notes. Search matches dn_number."""
    query = DeliveryNoteService.list_filters(
        workspace_id,
        status_filter,
        client_id,
        customer_purchase_order_id,
        invoice_id,
        search,
    )
    count_result = await session.execute(
        select(func.count()).select_from(query.subquery())
    )
    total = count_result.scalar()
    query = query.order_by(DeliveryNote.created_at.desc())
    query = query.offset((page - 1) * per_page).limit(per_page)
    result = await session.execute(query)
    rows = result.scalars().all()
    pages = (total + per_page - 1) // per_page if total else 0
    pagination = PaginationMeta(
        total=total,
        page=page,
        per_page=per_page,
        pages=pages,
        has_next=page < pages,
        has_prev=page > 1,
    )
    return PaginatedResponse(
        data=[DeliveryNoteService.serialize_list_item(row) for row in rows],
        pagination=pagination,
    )


@router.get("/{dn_id}", response_model=SuccessResponse[DeliveryNoteResponse])
async def get_delivery_note(
    dn_id: UUID,
    session: AsyncSession = Depends(get_session),
    workspace_id: UUID = Depends(get_current_workspace_id),
):
    """Get a delivery note with lines."""
    dn = await _load_or_404(session, dn_id, workspace_id)
    return await _wrapped(session, dn)


@router.put("/{dn_id}", response_model=SuccessResponse[DeliveryNoteResponse])
async def update_delivery_note(
    dn_id: UUID,
    body: DeliveryNoteUpdate,
    session: AsyncSession = Depends(get_session),
    user: User = Depends(get_current_user),
    workspace_id: UUID = Depends(get_current_workspace_id),
):
    """Replace DRAFT fields. Non-DRAFT → 403 INVALID_STATE."""
    dn = await _load_or_404(session, dn_id, workspace_id, for_update=True)
    patch = body.model_dump(exclude_unset=True)
    items = patch.pop("items", None)
    await DeliveryNoteService.update_draft(
        session=session, dn=dn, user_id=user.id, patch=patch, items=items
    )
    await session.commit()
    loaded = await DeliveryNoteService.get_visible(session, dn.id, workspace_id)
    return await _wrapped(session, loaded)


@router.delete("/{dn_id}", status_code=status.HTTP_200_OK)
async def delete_delivery_note(
    dn_id: UUID,
    session: AsyncSession = Depends(get_session),
    workspace_id: UUID = Depends(get_current_workspace_id),
):
    """Soft-delete a DRAFT delivery note. Does not rewind the number counter."""
    dn = await _load_or_404(session, dn_id, workspace_id, for_update=True)
    await DeliveryNoteService.soft_delete(dn)
    await session.commit()
    return SuccessResponse(
        data={"message": "Delivery note deleted successfully", "dn_id": str(dn_id)}
    )


@router.post(
    "/{dn_id}/confirm",
    response_model=SuccessResponse[DeliveryNoteResponse],
)
async def confirm_delivery_note(
    dn_id: UUID,
    _body: DeliveryNoteConfirmRequest = Body(
        default_factory=DeliveryNoteConfirmRequest
    ),
    session: AsyncSession = Depends(get_session),
    user: User = Depends(get_current_user),
    workspace_id: UUID = Depends(get_current_workspace_id),
):
    """DRAFT → CONFIRMED. Posts ISSUE for catalog lines. Idempotent 200."""
    dn = await _load_or_404(session, dn_id, workspace_id, for_update=True)
    await DeliveryNoteService.confirm(session, dn, user.id, workspace_id)
    await session.commit()
    loaded = await DeliveryNoteService.get_visible(session, dn.id, workspace_id)
    return await _wrapped(session, loaded)


@router.post(
    "/{dn_id}/cancel",
    response_model=SuccessResponse[DeliveryNoteResponse],
)
async def cancel_delivery_note(
    dn_id: UUID,
    body: DeliveryNoteCancelRequest = Body(default_factory=DeliveryNoteCancelRequest),
    session: AsyncSession = Depends(get_session),
    user: User = Depends(get_current_user),
    workspace_id: UUID = Depends(get_current_workspace_id),
):
    """CONFIRMED → CANCELLED. Reverses ISSUE stock."""
    dn = await _load_or_404(session, dn_id, workspace_id, for_update=True)
    await DeliveryNoteService.cancel(session, dn, user.id, body.reason)
    await session.commit()
    loaded = await DeliveryNoteService.get_visible(session, dn.id, workspace_id)
    return await _wrapped(session, loaded)

File: frontend/src/api/deliveryNotes.ts
import { apiClient } from './client';
import type { PaginatedResponse, PaginationMeta, SuccessResponse } from '../types/api';

export type DeliveryNoteStatus = 'DRAFT' | 'CONFIRMED' | 'CANCELLED';

export interface DeliveryNoteItemWrite {
  customer_purchase_order_item_id?: string;
  invoice_item_id?: string;
  quantity: number;
  bin_id?: string;
}

export interface DeliveryNoteItem {
  id: string;
  delivery_note_id: string;
  customer_purchase_order_item_id?: string | null;
  invoice_item_id?: string | null;
  product_id?: string | null;
  uom_id?: string | null;
  sku_snapshot?: string | null;
  description: string;
  quantity: number;
  bin_id?: string | null;
  created_at?: string;
  updated_at?: string;
}

export interface DeliveryNoteListItem {
  id: string;
  client_id: string;
  dn_number: string;
  status: DeliveryNoteStatus;
  delivery_date: string;
  customer_purchase_order_id?: string | null;
  invoice_id?: string | null;
  warehouse_id: string;
  created_at?: string;
}

export interface DeliveryNote extends DeliveryNoteListItem {
  workspace_id?: string;
  bin_id?: string | null;
  shipping_address?: string | null;
  vehicle_number?: string | null;
  driver_name?: string | null;
  notes?: string | null;
  cancellation_reason?: string | null;
  confirmed_by?: string | null;
  confirmed_at?: string | null;
  items: DeliveryNoteItem[];
  updated_at?: string;
  deleted_at?: string | null;
}

export interface DeliveryNoteCreatePayload {
  customer_purchase_order_id?: string;
  invoice_id?: string;
  warehouse_id: string;
  bin_id?: string;
  delivery_date?: string;
  shipping_address?: string;
  vehicle_number?: string;
  driver_name?: string;
  notes?: string;
  items?: DeliveryNoteItemWrite[];
}

export interface DeliveryNoteUpdatePayload {
  warehouse_id?: string;
  bin_id?: string | null;
  delivery_date?: string;
  shipping_address?: string | null;
  vehicle_number?: string | null;
  driver_name?: string | null;
  notes?: string;
  items?: DeliveryNoteItemWrite[];
}

export interface DeliveryNoteListQuery {
  page?: number;
  per_page?: number;
  status?: DeliveryNoteStatus;
  client_id?: string;
  customer_purchase_order_id?: string;
  invoice_id?: string;
  search?: string;
}

export interface DeliveryNoteListResult {
  items: DeliveryNoteListItem[];
  pagination: PaginationMeta;
}

const BASE = '/api/v1/delivery-notes';

function listParams(query: DeliveryNoteListQuery): Record<string, string | number> {
  const params: Record<string, string | number> = {
    page: query.page ?? 1,
    per_page: query.per_page ?? 20,
  };
  if (query.status) params.status = query.status;
  if (query.client_id) params.client_id = query.client_id;
  if (query.customer_purchase_order_id) {
    params.customer_purchase_order_id = query.customer_purchase_order_id;
  }
  if (query.invoice_id) params.invoice_id = query.invoice_id;
  if (query.search?.trim()) params.search = query.search.trim();
  return params;
}

function unwrapList(payload: PaginatedResponse<DeliveryNoteListItem>): DeliveryNoteListResult {
  const items = Array.isArray(payload.data) ? payload.data : [];
  return { items, pagination: payload.pagination };
}

function itemBody(item: DeliveryNoteItemWrite): Record<string, unknown> {
  const body: Record<string, unknown> = { quantity: item.quantity };
  if (item.customer_purchase_order_item_id) {
    body.customer_purchase_order_item_id = item.customer_purchase_order_item_id;
  }
  if (item.invoice_item_id) body.invoice_item_id = item.invoice_item_id;
  if (item.bin_id) body.bin_id = item.bin_id;
  return body;
}

function optionalText(value?: string | null): string | undefined {
  const trimmed = value?.trim();
  return trimmed || undefined;
}

export function createDeliveryNoteBody(
  data: DeliveryNoteCreatePayload,
): Record<string, unknown> {
  const body: Record<string, unknown> = { warehouse_id: data.warehouse_id };
  if (data.customer_purchase_order_id) {
    body.customer_purchase_order_id = data.customer_purchase_order_id;
  }
  if (data.invoice_id) body.invoice_id = data.invoice_id;
  if (data.bin_id) body.bin_id = data.bin_id;
  if (data.delivery_date) body.delivery_date = data.delivery_date;
  const address = optionalText(data.shipping_address);
  if (address) body.shipping_address = address;
  const vehicle = optionalText(data.vehicle_number);
  if (vehicle) body.vehicle_number = vehicle;
  const driver = optionalText(data.driver_name);
  if (driver) body.driver_name = driver;
  const notes = optionalText(data.notes);
  if (notes) body.notes = notes;
  if (data.items && data.items.length > 0) {
    body.items = data.items.map(itemBody);
  }
  return body;
}

export function updateDeliveryNoteBody(
  data: DeliveryNoteUpdatePayload,
): Record<string, unknown> {
  const body: Record<string, unknown> = {};
  if (data.warehouse_id) body.warehouse_id = data.warehouse_id;
  if (data.bin_id) body.bin_id = data.bin_id;
  if (data.delivery_date) body.delivery_date = data.delivery_date;
  if (data.shipping_address !== undefined) {
    body.shipping_address = optionalText(data.shipping_address) ?? null;
  }
  if (data.vehicle_number !== undefined) {
    body.vehicle_number = optionalText(data.vehicle_number) ?? null;
  }
  if (data.driver_name !== undefined) {
    body.driver_name = optionalText(data.driver_name) ?? null;
  }
  if (data.notes !== undefined) body.notes = optionalText(data.notes);
  if (data.items && data.items.length > 0) {
    body.items = data.items.map(itemBody);
  }
  return body;
}

export const getDeliveryNotes = async (
  query: DeliveryNoteListQuery = {},
): Promise<DeliveryNoteListResult> => {
  const response = await apiClient.get<PaginatedResponse<DeliveryNoteListItem>>(BASE, {
    params: listParams(query),
  });
  return unwrapList(response.data);
};

export const getDeliveryNote = async (id: string): Promise<DeliveryNote> => {
  const response = await apiClient.get<SuccessResponse<DeliveryNote>>(`${BASE}/${id}`);
  return response.data.data;
};

export const createDeliveryNote = async (
  data: DeliveryNoteCreatePayload,
): Promise<DeliveryNote> => {
  const response = await apiClient.post<SuccessResponse<DeliveryNote>>(
    BASE,
    createDeliveryNoteBody(data),
  );
  return response.data.data;
};

export const updateDeliveryNote = async (
  id: string,
  data: DeliveryNoteUpdatePayload,
): Promise<DeliveryNote> => {
  const response = await apiClient.put<SuccessResponse<DeliveryNote>>(
    `${BASE}/${id}`,
    updateDeliveryNoteBody(data),
  );
  return response.data.data;
};

export const deleteDeliveryNote = async (id: string): Promise<void> => {
  await apiClient.delete(`${BASE}/${id}`);
};

export const confirmDeliveryNote = async (id: string): Promise<DeliveryNote> => {
  const response = await apiClient.post<SuccessResponse<DeliveryNote>>(
    `${BASE}/${id}/confirm`,
    {},
  );
  return response.data.data;
};

export const cancelDeliveryNote = async (
  id: string,
  reason?: string,
): Promise<DeliveryNote> => {
  const body: { reason?: string } = {};
  if (reason) body.reason = reason;
  const response = await apiClient.post<SuccessResponse<DeliveryNote>>(
    `${BASE}/${id}/cancel`,
    body,
  );
  return response.data.data;
};
